src/worktree.rs

```diff
@@ -0,0 +1,327 @@
+    pub fn validate(&self) -> Result<(), Error> {
+        unsafe {
+            try_call!(raw::git_worktree_validate(self.raw));
+        }
+        Ok(())
+    }
+
+    /// Locks the worktree
+    pub fn lock(&self, reason: Option<&str>) -> Result<(), Error> {
+        let reason = crate::opt_cstr(reason)?;
+        unsafe {
+            try_call!(raw::git_worktree_lock(self.raw, reason));
+        }
+        Ok(())
+    }
+
+    /// Unlocks the worktree
+    pub fn unlock(&self) -> Result<(), Error> {
+        unsafe {
+            try_call!(raw::git_worktree_unlock(self.raw));
+        }
+        Ok(())
+    }
+
+    /// Checks if worktree is locked
+    pub fn is_locked(&self) -> Result<WorktreeLockStatus, Error> {
+        let buf = Buf::new();
+        unsafe {
+            match try_call!(raw::git_worktree_is_locked(buf.raw(), self.raw)) {
+                0 => Ok(WorktreeLockStatus::Unlocked),
+                _ => {
+                    let v = buf.to_vec();
+                    Ok(WorktreeLockStatus::Locked(match v.len() {
+                        0 => None,
+                        _ => Some(String::from_utf8(v).unwrap()),
+                    }))
+                }
+            }
+        }
+    }
+
+    /// Prunes the worktree
+    pub fn prune(&self, opts: Option<&mut WorktreePruneOptions>) -> Result<(), Error> {
+        // When successful the worktree should be removed however the backing structure
+        // of the git_worktree should still be valid.
+        unsafe {
+            try_call!(raw::git_worktree_prune(self.raw, opts.map(|o| o.raw())));
+        }
+        Ok(())
+    }
+
+    /// Checks if the worktree is prunable
+    pub fn is_prunable(&self, opts: Option<&mut WorktreePruneOptions>) -> Result<bool, Error> {
+        unsafe {
+            let rv = try_call!(raw::git_worktree_is_prunable(
+                self.raw,
+                opts.map(|o| o.raw())
+            ));
+            Ok(rv != 0)
+        }
+    }
+}
+
+impl<'a> WorktreeAddOptions<'a> {
+    /// Creates a default set of add options.
+    ///
+    /// By default this will not lock the worktree
+    pub fn new() -> WorktreeAddOptions<'a> {
+        unsafe {
+            let mut raw = mem::zeroed();
+            assert_eq!(
+                raw::git_worktree_add_options_init(&mut raw, raw::GIT_WORKTREE_ADD_OPTIONS_VERSION),
+                0
+            );
+            WorktreeAddOptions {
+                raw,
+                _marker: marker::PhantomData,
+            }
+        }
+    }
+
+    /// If enabled, this will cause the newly added worktree to be locked
+    pub fn lock(&mut self, enabled: bool) -> &mut WorktreeAddOptions<'a> {
+        self.raw.lock = enabled as c_int;
+        self
+    }
+
+    /// reference to use for the new worktree HEAD
+    pub fn reference(&mut self, reference: Option<&Reference<'a>>) -> &mut WorktreeAddOptions<'a> {
+        self.raw.reference = if let Some(reference) = reference {
```

Comment:
I'm not sure how safe this is in the sense that the `Reference` can be dropped before these options are used. I think that there needs to be a lifetime connection or this structure needs to take ownership of the `Reference`

Reply:
I've made sure that `Reference` will outlive `WorktreeAddOptions` by changing it to `pub fn reference(&mut self, reference: Option<&'a Reference<'_>>) -> &mut WorktreeAddOptions<'a>`

Reply:
Looks good to me, thanks!